### PR TITLE
Added plantuml for workflow description

### PR DESCRIPTION
**Pull request by rvyuha**
Added flow-for-automation.puml

### CODE REVIEW

File: specs/release-automation/flow-for-automation.puml [hunk 1]
```diff
@@ -0,0 +1,67 @@
+@startuml
+:User triggers the workflow;
+start
+:Create release files;
+while(unchecked release files)
+if (addHeader) then(true)
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:6]: Could you add the steps for actually creating the files? i.e. the CSV and Excel files?

File: specs/release-automation/flow-for-automation.puml [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+:Collapse headers;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:7]: What do you mean by `collapse headers`?

File: specs/release-automation/flow-for-automation.puml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+:Add new headers;
+endif
+if (release file on github) then(true)
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:10]: Shouldn't the deployment logic be done after creating the release files?

File: specs/release-automation/flow-for-automation.puml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+if(release version is less then version on github) then(true)
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:11]: Where are you getting the version number for what's on GitHub?

File: specs/release-automation/flow-for-automation.puml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+:Throw error;
+stop
+endif
+:Create new branch if one has not been made;
+:Create the release file;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:16]: Why is this here when there's already a "Create release files" step above?

File: specs/release-automation/flow-for-automation.puml [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+endif
+endwhile
+:Commit files to new branch;
+if(existing release branched/PRs) then(true)
+:Delete existing branches and PRs;
+endif
+:Push changes;
+:Create PR to dev with the new branch;
+stop
+:Release branch merged to main;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:31]: This should be done in a different workflow

File: specs/release-automation/flow-for-automation.puml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+if (release file on github) then(true)
+if(release version is less then version on github) then(true)
+:Throw error;
+stop
+endif
+:Create new branch if one has not been made;
+:Create the release file;
+:Add release file to correct location;
+else(false)
+if(release on OSF) then(true)
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:19]: Shouldn't thhis be done only if the PR is being merged to main

File: specs/release-automation/flow-for-automation.puml [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+:Create new directory for release;
+endif
+else(false)
+:Delete old files;
+endif
+endif
+endif
+:Upload OSF directory to OSF;
+stop
+:Updates to OSF and github are finished;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:51]: The next few steps don't have to go into depth. They can just say trigger a workflow in the PHES-ODM-doc and PHES-ODM-Validation repo.

File: specs/release-automation/flow-for-automation.puml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+:Commit files to new branch;
+if(existing release branched/PRs) then(true)
+:Delete existing branches and PRs;
+endif
+:Push changes;
+:Create PR to dev with the new branch;
+stop
+:Release branch merged to main;
+start
+if(release files found in OSF upload directory) then(true)
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:33]: What is the "OSF upload directory"?

File: specs/release-automation/flow-for-automation.puml [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+if(no release files on OSF) then(false)
+if(version on osf is different to current release version) then (true)
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:35]: How do you know what the version on OSF is?

File: specs/release-automation/flow-for-automation.puml (hunk 8 above)
Comment (yulric) [reply specs/release-automation/flow-for-automation.puml:51]: This should not be in a seperate axctivity diagram but should be done when the workflow is trigerred. I'll update specs.

File: specs/release-automation/flow-for-automation.puml [hunk 11]
```diff
@@ -0,0 +1,67 @@
+@startuml
+:User triggers the workflow;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:2]: There needs to be a step to validate the inputs and the files sheet.

**Review by yulric — CHANGES_REQUESTED** [review]

**Comment by yulric** [comment]
Another thing, can you also indent the file lines so its clear what lines are inside a while loop or if loop for example?

File: specs/release-automation/flow-for-automation.puml [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+group RA-1: Trigerring the Process
+    :User passes link to excel dictionary on OSF to use for the release;
+    :User passes OSF personal access token;
+    :User triggers release process;
+end group
+
+group RA-2: Creating the Release Files
+    :Retrieve Excel Dictionary;
+    :Parse metadata from files sheet;
+    :Construct file names, ending in ".csv" or ".xlsx";
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:14]: Can you expand on this step? I would like to know how your interpolation logic for a template string would work as well as the logic for retrieving the `version` template variable.

File: specs/release-automation/flow-for-automation.puml [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    :Populate file with information from appropriate sheet;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:15]: Can you provide more details on this step?

File: specs/release-automation/flow-for-automation.puml [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+    :Add optional headers;
+    :Store file in appropriate directory based on its upload location;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:17]: Should this be done here or in the deployment steps?

File: specs/release-automation/flow-for-automation.puml [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+end group
+
+group RA-3: Deploying the files to GitHub
+    :Check if previous release exists on GitHub in the appropriate directory;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:21]: We will need the `files` sheet used in the previous build for this right? If not, how were you planning on doing this?

File: specs/release-automation/flow-for-automation.puml [hunk 16]
```diff
@@ -0,0 +1,65 @@
...
+
+    if (No previous release on GitHub) then (yes)
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:23]: Try to avoid negatives in your if statements. For example, this one would be better read as, "Is there a previous release on GitHub?"

File: specs/release-automation/flow-for-automation.puml [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+        :Create new release branch and move files to the GitHub file storage directory;
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:24]: The old files need to be deleted

File: specs/release-automation/flow-for-automation.puml [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        if (Previous release is newer) then (yes)
+            :Stop process and throw an error;
+            stop
+        else (no)
+            :Move old files to archive directory and move new files to appropriate directory;
+        endif
+    endif
+    :Create PR from release branch to dev branch;
+end group
+group RA-4: Deploying the files to OSF
```
Comment (yulric) [inline specs/release-automation/flow-for-automation.puml:35]: I think this should be moved into a different workflow file one whose start is "Release branch merged to dev".

File: specs/release-automation/flow-for-automation.puml (hunk 16 above)
Comment (yulric) [reply specs/release-automation/flow-for-automation.puml:23]: Can you update all the other instances where this happens?

File: specs/release-automation/flow-for-automation.puml (hunk 18 above)
Comment (yulric) [reply specs/release-automation/flow-for-automation.puml:35]: So we should have two workflows files, one for the two trigger events below:

1. User triggers the workflow using the GitHub action tab
2. User merges a release branch into main

**Review by yulric — CHANGES_REQUESTED** [review]

**Comment by rvyuha** [comment]
@yulric Ready for review

File: specs/release-automation/user-triggered-automation-workflow.puml [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+    :User triggers release process;
+end group
+
+group RA-2: Creating the Release Files
+    :Retrieve Excel Dictionary;
+    :Parse metadata from files sheet;
+    :Extract file name and type;
+    :Extract dictionary version from the dictionary file name;
+    :Extract dictionary version from the dictionary summary sheet;
+    if (Version missmatch) then(yes)
```
Comment (yulric) [inline specs/release-automation/user-triggered-automation-workflow.puml:17]: Expand about version mismatch between what?

File: specs/release-automation/user-triggered-automation-workflow.puml [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+        :Issue error;
+        stop
+    else (no)
+        :Parse the name template string;
```
Comment (yulric) [inline specs/release-automation/user-triggered-automation-workflow.puml:21]: Lines 21 to 42 should go into a while loop right? which iterates through each row in the files sheet?

File: specs/release-automation/user-triggered-automation-workflow.puml [hunk 21]
```diff
@@ -0,0 +1,63 @@
...
+        :Construct file names using populated template and file type;
+        :Extract the part value from sheets for this file;
+        :Check parts sheet for the partID;
+        if (Part is a set) then (yes)
```
Comment (yulric) [inline specs/release-automation/user-triggered-automation-workflow.puml:25]: How do you know if a partID is a set from the parts sheet? Isn't it better to check the sets sheet?

File: specs/release-automation/user-triggered-automation-workflow.puml [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+            :Extract set information from sets with that part as ID;
+            :Extract sheets with the parts of the set;
+            if(File type is csv) then (yes)
+                :Issue warning;
```
Comment (yulric) [inline specs/release-automation/user-triggered-automation-workflow.puml:29]: We should record an error and stop the process if the creation process fails for any file.

File: specs/release-automation/user-triggered-automation-workflow.puml [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+            else (no)
+            endif
+        else (no)
+            :Extract sheets with that name;
+        endif
+        :Populate file with appropriate sheet(s);
+        if (Addheader is list) then (yes)
+            :Add optional headers;
+        endif
+        :Create github and OSF staging directories;
```
Comment (yulric) [inline specs/release-automation/user-triggered-automation-workflow.puml:40]: I don't think there's a need to create the OSF staging directories here, since we're not going to upload them.

File: specs/release-automation/user-triggered-automation-workflow.puml [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+    :Check if previous release exists on GitHub in the current upload destination;
+    :Create new branch from main following the "release-{version}" naming template;
+    if (Is there a previous release on GitHub) then (no)
+        :Move files from the staging github directory to their final location;
+    else (yes)
+        if (Previous release version number is newer) then (yes)
+            :Stop process and throw an error;
+            stop
+        else (no)
+            :Remove old files and move new files to appropriate directory;
```
Comment (yulric) [inline specs/release-automation/user-triggered-automation-workflow.puml:56]: How do you know where the old files are stored?

File: specs/release-automation/post-merge-automation-workflow.puml [hunk 25]
```diff
@@ -0,0 +1,28 @@
+@startuml
+title Post merge workflow
+:Dev branch release PR merged to main branch;
+start
+group RA-4: Deploying the files to OSF
+    if (Is there previous release on OSF) then (no)
```
Comment (yulric) [inline specs/release-automation/post-merge-automation-workflow.puml:6]: How do you know this?

File: specs/release-automation/post-merge-automation-workflow.puml [hunk 26]
```diff
@@ -0,0 +1,28 @@
...
+        :Push files in the OSF staging directory to OSF using the user's OSF key;
```
Comment (yulric) [inline specs/release-automation/post-merge-automation-workflow.puml:7]: Did you investigate if you can store files in an action?

**Review by yulric — CHANGES_REQUESTED** [review]

File: specs/release-automation/post-merge-automation-workflow.puml (hunk 26 above)
Comment (rvyuha) [reply specs/release-automation/post-merge-automation-workflow.puml:7]: Yes and they can be stored and access across different jobs. And even non github action jobs can access them.

File: specs/release-automation/user-triggered-automation-workflow.puml [hunk 27]
```diff
@@ -0,0 +1,76 @@
...
+    if (Version missmatch between dictionary file name version and summary sheet version?) then(yes)
+        :Issue error;
+        stop
+    else (no)
+        repeat
+            :Parse the name template string;
+            :Construct file name using populated template and file type;
+            :Extract the part value from sheets for this file;
+            :Check parts sheet for the partID;
+            if (Part is a partType set and has matching information in the sets sheet?) then (yes)
```
Comment (yulric) [inline specs/release-automation/user-triggered-automation-workflow.puml]: I don't think there's a `partType` called `set` in the dictionary?

**Review by yulric — CHANGES_REQUESTED** [review]
Can you add what we discussed about how to get the old dictionary from OSF?

**Review by yulric — APPROVED** [review]
